src/ray/object_manager/pull_manager.cc

```diff
@@ -40,27 +40,31 @@ void PullManager::OnLocationChange(const ObjectID &object_id,
   if (it == pull_requests_.end()) {
     return;
   }
-  // Reset the list of clients that are now expected to have the object.
-  // NOTE(swang): Since we are overwriting the previous list of clients,
-  // we may end up sending a duplicate request to the same client as
-  // before.
-  it->second.client_locations = std::vector<NodeID>(client_ids.begin(), client_ids.end());
-  if (!spilled_url.empty()) {
-    RAY_LOG(DEBUG) << "OnLocationChange " << spilled_url << " num clients "
-                   << client_ids.size();
-    // Try to restore the spilled object.
-    restore_spilled_object_(object_id, spilled_url,
-                            [this, object_id](const ray::Status &status) {
-                              // Fall back to fetching from another object manager.
-                              if (!status.ok()) {
-                                TryPull(object_id);
-                              }
-                            });
-  } else {
-    // New object locations were found, so begin trying to pull from a
-    // client. This will be called every time a new client location
-    // appears.
-    TryPull(object_id);
+  auto &request = it->second;
+  if (!request.inflight_pull) {
```

Comment:
Hmm this doesn't seem like it will work well for object broadcast. Won't all of the workers just try to request from the first location only?

Reply:
Updated, though I depending on the push manager/network scheduling policy this may not be great broadcast performance still (not a regression, just not great). ff